src/managed/Microsoft.DotNet.Build.Bundle/BundleManifest.cs

```diff
@@ -0,0 +1,208 @@
+    /// -----------------------------------------------
+    /// Meta-data Directory (FileEntries)
+    ///   "app.deps.json"               (if any)
+    ///   "app.runtimeconfig.json"      (if any)
+    ///   "app.runtimeconfig.dev.json"  (if any)
+    ///   "app.dll"                 (The main app)
+    ///   "IL_1.dll"                (Assemblies, if any)
+    ///    ...
+    ///   "IL_n.dll"
+    ///   "File_1"                  (Other Files, if any)
+    ///    ...
+    ///   "File_n" 
+    /// - - - - - - - - - - - - - - - - - - - - - - - -
+    ///  Bundle Manifest (footer)
+    ///     MajorVersion
+    ///     MinorVersion
+    ///     FileEntryStart (offset)
+    ///     NumFiles
+    ///     NumAssemblies
+    ///     Flags
+    /// -----------------------------------------------
+    ///   Bundle Signature
+    /// _________________________________________________
+    /// 
+    /// The Bundle manifest does not encode FileType explicitly in each entry. 
+    /// The FileType for each embedded file is deduced based on:
+    ///  - FileEntry position: FileEntries are recorded in the order
+    ///     -- Configuration files
+    ///     -- The main assembly
+    ///     -- Other assemblies
+    ///     -- Other Files
+    ///  - Flags (to determine number of configuration files)
+    ///  - NumAssemblies 
+    /// </summary>
+
+    public class BundleManifest
+    {
+        [Flags]
+        enum BundleFlags
+        {
+            None = 0,
+            HasDepsJson = 1,
+            HasRuntimeConfigJson = 2,
+            HasRuntimeConfigDevJson = 4
+        }
+
+        public const string Signature = ".NetCoreBundle";
+        public const uint MajorVersion = 0;
+        public const uint MinorVersion = 1;
+
+        const uint SignatureLength = 15;
+        const uint FooterLength = 43;
+
+        long FileEntryStart;
+        int AssemblyCount;
+        BundleFlags flags;
+
+        public List<FileEntry> Files;
+
+        public BundleManifest()
+        {
+            Files = new List<FileEntry>();
+            flags = BundleFlags.None;
+        }
+
+        public FileEntry AddEntry(FileType type, string name, long offset, long size)
+        {
+            switch(type)
+            {
+                case FileType.DepsJson:
+                    flags |= BundleFlags.HasDepsJson;
+                    break;
+                case FileType.RuntimeConfigJson:
+                    flags |= BundleFlags.HasRuntimeConfigJson;
+                    break;
+                case FileType.RuntimeConfigDevJson:
+                    flags |= BundleFlags.HasRuntimeConfigDevJson;
+                    break;
+
+                default:
+                    break;
+            }
+
+            FileEntry entry = new FileEntry(type, name, offset, size);
+            Files.Add(entry);
+            return entry;
+        }
+
+        void WriteFileEntries(BinaryWriter writer)
+        {
+            Action<FileType> writeEntry = (FileType type) =>
```

Comment:
You could use local functions for this - would look nicer (the end result is exactly the same, just different syntax)